javascripts/background_scripts/modal_button.js

```diff
@@ -1,64 +1,110 @@
- *
+ * @fileOverview This file initializes the modal button (browser action).
+ * This file initializing the button's state on browser start and
+ * change button's state (badage color and text) when user toggles
+ * injection in popup.html.
+ * 
  * The text message on the button determines the operating mode.
  * "off" indicates that the content script can be injected, but it
  * doesn't execute any code.
  *
  * "on" indicates that the content script is injected, and should
  * actively replace links on the user's whitelist.
- *
  */
 
-
-  /*global chrome:false, ls:true, readingProcess:false, */
+/*global chrome */
+/*global Privly */
 
 /**
- * @namespace for the modal button.
+ * This function updates the badge text and background color of the
+ * browser action button (modal button) according to the parameter.
+ * 
+ * @param  {Boolean} enableInjection Whether injection is enabled
  */
-var modalButton = {
+var injectionEnabled = Privly.options.isInjectionEnabled();;
 
-  /**
-   * Gives the current state of the modal button. This is provided so that
-   * an assynchronous call to getBadgeText is not necessary.
-   */
-  badgeText: "on",
+function updateBrowserAction() {
+  if (isInEmbededPosting()) {
+    chrome.browserAction.setIcon({
+      path: {
+        19: 'images/icon_writing_19.png',
+        38: 'images/icon_writing_38.png',
+      }
+    });
+  } else {
+    if (injectionEnabled) {
+      chrome.browserAction.setIcon({
+        path: {
+          19: 'images/icon_enabled_19.png',
+          38: 'images/icon_enabled_38.png',
+        }
+      });
+    } else {
+      chrome.browserAction.setIcon({
+        path: {
+          19: 'images/icon_disabled_19.png',
+          38: 'images/icon_disabled_38.png',
+        }
+      });
+    }
+  }
+}
 
+function isInEmbededPosting() {
+  if (embededStatus[currentActiveTabId] === undefined) {
+    return false;
+  }
+  if (Object.keys(embededStatus[currentActiveTabId]).length === 0) {
+    return false;
+  }
+  return true;
+}
 
-  /**
-   * Handles when the popup.js script sends a command to change the
-   * operating mode of the content script. This message handler
-   * is selected with a message "modeChange".
-   *
-   * @param {object} request An object containing the message body.
-   * @param {object} sender The scripting context that sent the message.
-   * @param {function} sendResponse The sender can send a function that
-   * will return a message.
-   */
-  modeChange: function(request, sender, sendResponse) {
-    if( request.handler === "modeChange" ) {
-      if( modalButton.badgeText === "on") {
-        modalButton.badgeText = "off";
-      } else {
-        modalButton.badgeText = "on";
-      }
-      chrome.tabs.query({currentWindow:true, highlighted: true}, readingProcess.tabsChange);
-      sendResponse();
+// Subscribe to option changed events
+Privly.message.addListener(function (request, sendRequest, sender) {
+  if (request.action === 'options/changed') {
```

Comment:
This should be declared in a separate function that is added as a listener so you can document/comment it and unit test it. This is true several other places as well.